### PR TITLE
Added commit message style check

### LINKED ISSUE #20: [qa] Add commit message style check
It's time to automate the commit message style checks according to [our guidelines](http://openwisp.io/docs/developer/contributing.html#commit-message-style-guidelines), that will make contributing a lot smoother for new contributors and less time consuming for reviewers.

These are the most common issues which we should address:

- prefix not present, eg: `Done something` (should be ``[qa] Done something``)
- no capital letter after prefix: `[qa] done something`
- unnecessary final dot in first line, eg: `[qa] Done something.`
- No blank line before first line and second line, eg:
```
[qa] Done something
Long description
```
- Mentions an issue without closing it, eg: `Implements #20` (won't close), `Closes #20 and #21` (won't close `#21` - but we should allow writing something like `Related to #33`, because sometimes we may want to reference an issue even though we are not solving it in that commit
- Closes an issue in the long description but doesn't mention it in the short description, eg:
```
[qa] Done something

Fixes #2
```
should be:
```
[qa] Done something #2

Fixes #2
```
(easier to read in the short log format)

This is not urgent but not unimportant because having a clean commit history is important for the long term sustainability of the project.
It doesn't make sense to keep enforcing these rules by hand.. it's clear to me that as the community grows we won't be able to do that, so automating this check is key.

We need something like the migration check implemented in #16.

### PR DESCRIPTION
**Pull request by ppabcd**
Google Code-in Task : https://codein.withgoogle.com/dashboard/task-instances/5188396768034816/
- [x] prefix not present, eg: Done something (should be [qa] Done something)
- [x] no capital letter after prefix: [qa] done something
- [x] unnecessary final dot in first line, eg: [qa] Done something.
- [x] No blank line before first line and second line
- [x] Mentions an issue without closing it
- [x] Closes an issue in the long description but doesn't mention it in the short description

Fixes #20

### CODE REVIEW

**Comment by ppabcd** [comment]
ok i'll try to move my code.

**Comment by KylaCommit** [comment]
You have errors with commit message: 
- No mention issues in the short description

**Comment by atb00ker** [comment]
@nemesisdesign @ppabcd 
I think an easier way to solve this problem is using the travis environment variables: `TRAVIS_COMMIT` and `TRAVIS_COMMIT_MESSAGE` and preforming simple regex tests.
If the regex expression doesn't match, we can simply fail the build and print a message there.

I feel this is a bit complicated way to solve this problem. What do you guys think?

**Comment by nemesisdesign** [comment]
@atb00ker sounds good.

Let's keep in mind we'll have to implement this check in `.travis.yml` using a trick like this one:
https://denysdovhan.com/building-pull-requests-using-travis/

We want this check to run only on pull-requests, because once a commit is done on master, with the new branch protection rules in place we can't change it (that means admins will have to be forced to use more and more development branches and avoid working on master, which is a good thing).

File: setup.py [hunk 1]
```diff
@@ -33,7 +33,8 @@
     packages=find_packages(exclude=['tests', 'docs']),
     entry_points={
         'console_scripts': [
-            'checkmigrations = openwisp_utils.qa:check_migration_name'
+            'checkmigrations = openwisp_utils.qa:check_migration_name',
+            'commitcheck = openwisp_utils.qa:check_commit_message',
```
Comment (nemesisdesign) [inline setup.py]: use `checkcommit` for consistency

File: tests/test_project/tests/test_qa.py [hunk 2]
```diff
@@ -6,13 +6,16 @@
...
+from openwisp_utils.qa import check_commit_message, check_migration_name
 
 MIGRATIONS_DIR = path.join(path.dirname(path.dirname(path.abspath(__file__))), 'migrations')
 
 
 class TestQa(TestCase):
     _test_migration_file = '%s/0002_auto_20181001_0421.py' % MIGRATIONS_DIR
+    _gh_token = os.environ["GH_TOKEN"]
+    _repo = os.environ["TRAVIS_REPO_SLUG"]
+    _commit = os.environ["TRAVIS_COMMIT"]
```
Comment (nemesisdesign) [inline tests/test_project/tests/test_qa.py]: I don't think you need to do this here, you need to supply your own test values directly

File: tests/test_project/tests/test_qa.py [hunk 3]
```diff
@@ -44,5 +47,35 @@ def test_qa_call_check_migration_name_failure(self):
                 else:
                     self.fail('SystemExit or Exception not raised')
 
+    def test_qa_call_check_commit_message_pass(self):
+        options = [
+            'commitcheck', '--token', self._gh_token,
+            '--commit', self._commit,
```
Comment (nemesisdesign) [inline tests/test_project/tests/test_qa.py]: it's ok to hardcode values here, you should test all the pass cases that we mentioned in the discussion

File: tests/test_project/tests/test_qa.py [hunk 4]
```diff
@@ -44,5 +47,35 @@ def test_qa_call_check_migration_name_failure(self):
...
+            '--repo', self._repo
+        ]
+        with patch('argparse._sys.argv', options):
+            check_commit_message()
+
+    def test_qa_call_check_commit_message_failure(self):
+        options = [
+            [
+                'commitcheck', '--token', self._gh_token,
+                '--commit', self._commit,
```
Comment (nemesisdesign) [inline tests/test_project/tests/test_qa.py]: it's ok to hardcode values here, you should test all the failure cases that we mentioned in the discussion

File: openwisp_utils/qa.py [hunk 5]
```diff
@@ -51,3 +53,130 @@ def check_migration_name():
...
+                        help="Repository target <username>/<repo>")
+    return parser.parse_args()
+
+
+def check_commit_message():
+    args = _commit_check_args()
+
+    if not args.token:
+        if "GH_TOKEN" in os.environ:
+            github_token = os.environ["GH_TOKEN"]
```
Comment (ppabcd) [inline openwisp_utils/qa.py]: I don't know how to coverage this line in travis

File: tests/test_project/tests/test_qa.py [hunk 6]
```diff
@@ -44,5 +44,70 @@ def test_qa_call_check_migration_name_failure(self):
                 else:
                     self.fail('SystemExit or Exception not raised')
 
+    def test_qa_call_check_commit_message_pass(self):
+        options = [
+            'commitcheck', '--token', os.environ["GH_TOKEN"],
+            '--commit', os.environ["TRAVIS_COMMIT"],
```
[the remark this answers is not in the document]
Comment (ppabcd) [reply tests/test_project/tests/test_qa.py]: commit hash

**Comment by KylaCommit** [comment]
You have errors with commit message: 
- Do not add a final dot at the end of the first line.
- No mention issues in the short description
- No blank line before first line and second line

**Comment by KylaCommit** [comment]
You have errors with commit message: 
- Do not add a final dot at the end of the first line.
- No mention issues in the short description
- No blank line before first line and second line

**Comment by KylaCommit** [comment]
All check done, no errors.

**Comment by ppabcd** [comment]
I get more problem : 
- In travis-ci.org i can't access commit title. You can see different [here](https://travis-ci.org/openwisp/openwisp-utils/jobs/458920966) with [my travis](https://travis-ci.com/ppabcd/openwisp-utils/jobs/160560448)
- I can't create encrypted `GH_TOKEN` and now i disable comment notification and only return travis error.
- ` No blank line before first line and second line` I see when you added title *Test* with description *Testing Commit*. Github will create commit messages like this.
```
Testing

Testing Commit
```
is you mean first line and second line in description?
- I am currently looking for an algorithm `Mentions an issue without closing it`. Because i don't have much experience in regex.

File: openwisp_utils/qa.py [hunk 7]
```diff
@@ -51,3 +53,166 @@ def check_migration_name():
...
+def check_commit_message():
+    args = _commit_check_args()
+
+    commit = _get_commit()[0]
+
+    # Check args
+    # if not args.token:
+    #     raise Exception("CLI arguement `token` is required "
+    #                     "but not found")
+    if not args.repo:
```
[the remark this answers is not in the document]
Comment (ppabcd) [reply openwisp_utils/qa.py]: this argument will be deleted

File: openwisp_utils/qa.py [hunk 8]
```diff
@@ -51,3 +53,166 @@ def check_migration_name():
...
+
+    if args.message:
+        api_title = args.message.split("\n")[0]
+        if len(args.message.split("\n")) > 1:
+            api_message = args.message.split("\n")[1:]
+    else:
+        api_title = commit['title']
+
+        if commit['message'] != '':
+            api_message = commit['message'].split("\n")
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: rename all variables that have the `api_` prefix so that they don't have it anymore

File: openwisp_utils/qa.py [hunk 9]
```diff
@@ -51,3 +53,166 @@ def check_migration_name():
...
+    prefix = re.match(r'\[(.*?)\]', api_title)
+    print(prefix)
+    if not prefix:
+        errors.append(
+            "Add prefix in beginning of the commit. "
+            "Example: [module/file/feature]"
+        )
+    else:
+        # Check capital after prefix
+        commitMessagefirst = api_title.replace(prefix.group(), '').strip()
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: do not use camel case variables in python

File: openwisp_utils/qa.py [hunk 10]
```diff
@@ -51,3 +53,166 @@ def check_migration_name():
...
+
+
+def _shell_exec(cmd):
+    return os.popen(cmd).read()
+
+
+# Thank you https://gist.github.com/simonw/091b765a071d1558464371042db3b959
+def _get_commit():
+    leading_4_spaces = re.compile('^    ')
+    lines = _shell_exec('git log -1').split('\n')
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: I don't think you need to do this, we'll pass output of git log via the bash shell

File: openwisp_utils/qa.py [hunk 11]
```diff
@@ -51,3 +53,166 @@ def check_migration_name():
...
+        raise Exception("CLI arguement `repo` is required "
+                        "but not found")
+    # Init var
+    # api = "https://api.github.com"
+
+    # If using token
+    # api_token = args.token
+
+    if args.message:
+        api_title = args.message.split("\n")[0]
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: is `api_title` the commit short desc?

File: tests/test_project/tests/test_qa.py [hunk 12]
```diff
@@ -44,5 +44,66 @@ def test_qa_call_check_migration_name_failure(self):
                 else:
                     self.fail('SystemExit or Exception not raised')
 
+    def test_qa_call_check_commit_message_pass(self):
+        options = [
+            'commitcheck',
+            # '--token',
+            # os.environ["GH_TOKEN"],
+            '--repo', os.environ["TRAVIS_REPO_SLUG"],
+            '--message', os.environ['TRAVIS_COMMIT_MESSAGE']
```
[the remark this answers is not in the document]
Comment (nemesisdesign) [reply tests/test_project/tests/test_qa.py]: These are all good commit messages, use these ones as starting point

```
[qa] Minor clean up operations
```
(it's ok to not reference any issue sometimes)

```
[qa] Improved Y #2

Related to #2
```
(but if an issue is mentioned it should be specified whether the commit fixes it or just references it)

```
[qa] Finished task #2

Closes #2
Related to #1
```
(multiple references can be added in the extended desc)

These commit messages must not fail.

File: tests/test_project/tests/test_qa.py [hunk 13]
```diff
@@ -44,5 +44,66 @@ def test_qa_call_check_migration_name_failure(self):
...
+            ],
+            [
+                'commitcheck',
+                # '--token',
+                # os.environ["GH_TOKEN"],
+                '--message',
+                '[qa] Hello World.\nTest\nFixes #20',
+                '--repo', "ppabcd/openwisp-utils"
+            ],
+            ['commitcheck']
```
Comment (nemesisdesign) [inline tests/test_project/tests/test_qa.py]: other common bad commit messages you should add to the failing tests:

```
[qa] Finished task X #2
Closes #2
```
no blank line between short desc and extended desc

File: openwisp_utils/qa.py (hunk 11 above)
Comment (ppabcd) [reply openwisp_utils/qa.py]: Yup

File: openwisp_utils/qa.py (hunk 9 above)
Comment (ppabcd) [reply openwisp_utils/qa.py]: i'll change

File: openwisp_utils/qa.py [hunk 14]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+
+def _commit_check_args():
+    """
+    Parse and return CLI arguements
+    """
+    parser = argparse.ArgumentParser(description="Ensures migration files "
+                                     "created have a descriptive name. If "
+                                     "default name pattern is found, "
+                                     "raise exception!")
+    parser.add_argument("--token",
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: shouldn't you remove this?

File: openwisp_utils/qa.py [hunk 15]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+                        help="Github token for check commit")
+    parser.add_argument("--message",
+                        help="Commit message")
+    parser.add_argument("--repo",
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: shouldn't you remove this?

File: openwisp_utils/qa.py (hunk 11 above)
Comment (nemesisdesign) [reply openwisp_utils/qa.py]: change all these variable names so they don't mention api anymore, if this is the short desc, name it `short_desc`, apply the same suggestion to all variables

File: openwisp_utils/qa.py [hunk 16]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+    args = _commit_check_args()
+
+    if not args.message:
+        raise Exception("CLI arguement `message` is required "
+                        "but not found")
+    else:
+        api_title = args.message.split("\n")[0]
+        if len(args.message.split("\n")) > 1:
+            api_message = args.message.split("\n")[1:]
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: openwisp_utils/qa.py [hunk 17]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+    errors = []
+    # Check dot in end of the first line
+    if api_title[len(api_title.strip()) - 1].strip() == '.':
+        errors.append("Do not add a final dot at the end of the first line.")
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: openwisp_utils/qa.py (hunk 9 above)
Comment (nemesisdesign) [reply openwisp_utils/qa.py]: address this comment

File: openwisp_utils/qa.py [hunk 18]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+        errors.append(
+            "Add prefix in beginning of the commit. "
+            "Example: [module/file/feature]"
+        )
+    else:
+        # Check capital after prefix
+        commitMessagefirst = api_title.replace(prefix.group(), '').strip()
+        if not commitMessagefirst[0].isupper():
+            errors.append("No capital letter after prefix")
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: openwisp_utils/qa.py [hunk 19]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+    # Check hastag
+    hastag = re.search(r'\#(\d+)', api_title)
+    if not hastag:
+        errors.append("No mention issues in the short description")
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: openwisp_utils/qa.py [hunk 20]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+    # Check blank page before first line and second line
+    if 'api_message' in locals():
+        if len(args.message.split("\n")) > 1:
+            if args.message.split("\n")[1] != '':
+                errors.append("No blank line before first "
+                              "line and second line")
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: openwisp_utils/qa.py [hunk 21]
```diff
@@ -51,3 +51,94 @@ def check_migration_name():
...
+                'refers to',
+                'ref',
+                'rel'
+            ]
+            parts = message.split(' ')
+            num = len(parts)
+            word = parts[num - 2]
+            if word.lower() not in allowed_words:
+                errors.append("Mentions an issue without closing it")
+
```
Comment (nemesisdesign) [inline openwisp_utils/qa.py]: remove blank line

File: tests/test_project/tests/test_qa.py [hunk 22]
```diff
@@ -44,5 +44,54 @@ def test_qa_call_check_migration_name_failure(self):
                 else:
                     self.fail('SystemExit or Exception not raised')
 
+    def test_qa_call_check_commit_message_pass(self):
+        options = [
+            'commitcheck',
+            '--message',
+            "[qa] Updated more file and fix "
+            "problem #20\n\nAdded more files Fixes #20"
```
Comment (nemesisdesign) [inline tests/test_project/tests/test_qa.py]: These are all good commit messages, use these ones as starting point

```
[qa] Minor clean up operations
```
(it's ok to not reference any issue sometimes)

```
[qa] Improved Y #2

Related to #2
```
(but if an issue is mentioned it should be specified whether the commit fixes it or just references it)

```
[qa] Finished task #2

Closes #2
Related to #1
```
(multiple references can be added in the extended desc)

These commit messages must not fail.

**Comment by ppabcd** [comment]
ready @nemesisdesign

File: openwisp_utils/qa.py [hunk 23]
```diff
@@ -51,3 +51,147 @@ def check_migration_name():
         raise Exception("Migration %s %s in directory %s must "
                         "be renamed to something more descriptive."
                         % (file_, ', '.join(migrations), args.migration_path))
+
+
+def _commit_check_args():
+    """
+    Parse and return CLI arguments
+    """
+    parser = argparse.ArgumentParser(description="Ensures migration files "
```
Comment (atb00ker) [inline openwisp_utils/qa.py]: The description of this argument seems like it's been copied from above! Please check this!

Comment (atb00ker) [reply openwisp_utils/qa.py]:
```
description="Ensures the commit message "
"matches the OpenWISP commit guidelines.")
```

**Comment by ppabcd** [comment]
i have done updated description messages @atb00ker

**Comment by ppabcd** [comment]
all done @nemesisdesign @atb00ker

**Pull request merged**